File: Java/app/src/main/java/kz/sgq/fs_tiktaktoe/mvp/model/FieldModelImpl.java
package kz.sgq.fs_tiktaktoe.mvp.model;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import kz.sgq.fs_tiktaktoe.mvp.model.interfaces.FieldModel;
import kz.sgq.fs_tiktaktoe.utilit.AnswerField;

public class FieldModelImpl implements FieldModel {

    private String playerOne;
    private String playerTwo;
    private int scorePlayerOne;
    private int scorePlayerTwo;
    private List<Integer> listBoxOne;
    private List<Integer> listBoxTwo;
    private List<Integer> boxWinList;
    private boolean step = true;
    private boolean end = false;
    private String stepText;

    public FieldModelImpl(String playerOne, String playerTwo) {
        this.playerOne = playerOne;
        this.playerTwo = playerTwo;
        scorePlayerOne = 0;
        scorePlayerTwo = 0;
        createLists();
    }

    @Override
    public String getScorePlayerOne() {
        return playerOne + " " + scorePlayerOne;
    }

    @Override
    public String getScorePlayerTwo() {
        return scorePlayerTwo + " " + playerTwo;
    }

    @Override
    public String getStep() {
        if (step)
            return stepText + " " + playerOne;
        else
            return stepText + " " + playerTwo;
    }

    @Override
    public boolean onClickBox(int i) {
        if (step) {
            if (listBoxOne.get(i) == -1 &&
                    listBoxTwo.get(i) == -1) {
                listBoxOne.set(i, i);
                return true;
            }
        } else {
            if (listBoxOne.get(i) == -1 &&
                    listBoxTwo.get(i) == -1) {
                listBoxTwo.set(i, i);
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean isStep() {
        return step;
    }

    @Override
    public boolean checkVictory() {
        if (isStep()) {
            if (checkBox(listBoxOne)) {
                scorePlayerOne++;
                end = true;
                return true;
            } else
                return false;
        } else {
            if (checkBox(listBoxTwo)) {
                scorePlayerTwo++;
                end = true;
                return true;
            } else
                return false;
        }
    }

    @Override
    public void setStep(boolean b) {
        step = b;
    }

    @Override
    public void createLists() {
        boxWinList = new ArrayList<>();
        listBoxOne = new ArrayList<>();
        listBoxTwo = new ArrayList<>();
        for (int i = 0; i < 9; i++) {
            listBoxOne.add(-1);
            listBoxTwo.add(-1);
        }
        end = false;
    }

    @Override
    public boolean checkDraw() {
        for (int i = 0; i < 8; i++) {
            if (listBoxOne.get(i) == -1 &&
                    listBoxTwo.get(i) == -1)
                return false;
        }
        return true;
    }

    @Override
    public void clearScore() {
        scorePlayerOne = 0;
        scorePlayerTwo = 0;
    }

    @Override
    public void setStepText(String str) {
        stepText = str;
    }

    @Override
    public int stepBot() {
        int checkStepWin = checkStepBot(listBoxTwo);
        if (checkStepWin == -1) {
            int checkStepEnemy = checkStepBot(listBoxOne);
            if (checkStepEnemy == -1) {
                List<Integer> list = stepRandomBot();
                Random random = new Random(System.currentTimeMillis());
                return list.get(random.nextInt(list.size()));
            } else
                return checkStepEnemy;
        } else
            return checkStepWin;
    }

    @Override
    public boolean getEnd() {
        return end;
    }

    @Override
    public List<Integer> getBoxWinList() {
        return boxWinList;
    }

    private int checkStepBot(List<Integer> list) {
        for (int i = 0; i < 8; i++) {
            boolean boxOne = false;
            boolean boxTwo = false;
            boolean boxThree = false;
            for (int j = 0; j < list.size(); j++) {
                if (AnswerField.getAnswer(i, 0) == list.get(j)) boxOne = true;
                if (AnswerField.getAnswer(i, 1) == list.get(j)) boxTwo = true;
                if (AnswerField.getAnswer(i, 2) == list.get(j)) boxThree = true;
            }
            int answerOne = AnswerField.getAnswer(i, 2);
            int answerTwo = AnswerField.getAnswer(i, 1);
            int answerThree = AnswerField.getAnswer(i, 0);
            if (boxOne && boxTwo &&
                    listBoxOne.get(answerOne) == -1 &&
                    listBoxTwo.get(answerOne) == -1)
                return answerOne;
            if (boxOne && boxThree &&
                    listBoxOne.get(answerTwo) == -1 &&
                    listBoxTwo.get(answerTwo) == -1)
                return answerTwo;
            if (boxTwo && boxThree &&
                    listBoxOne.get(answerThree) == -1 &&
                    listBoxTwo.get(answerThree) == -1)
                return answerThree;
        }
        return -1;
    }

    private List<Integer> stepRandomBot() {
        List<Integer> list = new ArrayList<>();
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < listBoxOne.size(); j++) {
                int answerOne = AnswerField.getAnswer(i, 0);
                int answerTwo = AnswerField.getAnswer(i, 1);
                int answerThree = AnswerField.getAnswer(i, 2);
                if (answerOne != listBoxOne.get(j) &&
                        listBoxOne.get(answerOne) == -1 &&
                        listBoxTwo.get(answerOne) == -1) {
                    list.add(answerOne);
                }
                if (answerTwo != listBoxOne.get(j) &&
                        listBoxOne.get(answerTwo) == -1 &&
                        listBoxTwo.get(answerTwo) == -1) {
                    list.add(answerTwo);
                }
                if (answerThree != listBoxOne.get(j) &&
                        listBoxOne.get(answerThree) == -1 &&
                        listBoxTwo.get(answerThree) == -1) {
                    list.add(answerThree);
                }
            }
        }
        return list;
    }

    private boolean checkBox(List<Integer> list) {
        for (int i = 0; i < 8; i++) {
            boolean boxOne = false;
            boolean boxTwo = false;
            boolean boxThree = false;
            for (int j = 0; j < list.size(); j++) {
                if (AnswerField.getAnswer(i, 0) == list.get(j)) boxOne = true;
                if (AnswerField.getAnswer(i, 1) == list.get(j)) boxTwo = true;
                if (AnswerField.getAnswer(i, 2) == list.get(j)) boxThree = true;
            }
            if (boxOne && boxTwo && boxThree) {
                boxWinList.add(AnswerField.getAnswer(i, 0));
                boxWinList.add(AnswerField.getAnswer(i, 1));
                boxWinList.add(AnswerField.getAnswer(i, 2));
                return true;
            }
        }
        return false;
    }
}

File: Kotlin/app/src/main/java/kz/sgq/kotlintictactoe/mvp/model/FieldModelImpl.kt
package kz.sgq.kotlintictactoe.mvp.model

import kz.sgq.kotlintictactoe.mvp.model.interfaces.FieldModel
import kz.sgq.kotlintictactoe.utils.AnswerField
import java.util.*

const val ANSWER_NULL: Int = -1

class FieldModelImpl(
        var playerOne: String,
        var playerTwo: String
) : FieldModel {

    private var stepText: String? = null
    private var scorePlayerOne: Int = 0
    private var scorePlayerTwo: Int = 0
    private var step: Boolean = true
    private var mode: Boolean = true
    private var end: Boolean = false
    private var listBoxOne: MutableList<Int> = arrayListOf()
    private var listBoxTwo: MutableList<Int> = arrayListOf()
    private var boxWinList: MutableList<Int> = arrayListOf()

    init {
        createLists()
    }

    override fun getScorePlayerOne(): String = "$playerOne $scorePlayerOne"

    override fun getScorePlayerTwo(): String = "$scorePlayerTwo $playerTwo"

    override fun getStep(): String = if (step) "$stepText $playerOne"
    else "$stepText $playerTwo"

    override fun onClickBox(i: Int): Boolean {
        if (listBoxOne[i] == ANSWER_NULL &&
                listBoxTwo[i] == ANSWER_NULL) {
            if (step)
                listBoxOne[i] = i
            else
                listBoxTwo[i] = i
            return true
        }
        return false
    }

    override fun isStep(): Boolean = step

    override fun isMode(): Boolean {
        return mode
    }

    override fun setMode(mode: Boolean) {
        this.mode = mode
    }
    override fun checkVictory(): Boolean {
        return if (step)
            if (checkBox(listBoxOne)) {
                scorePlayerOne++
                end = true
                true
            } else
                false
        else
            if (checkBox(listBoxTwo)) {
                scorePlayerTwo++
                end = true
                true
            } else
                false
    }

    override fun checkDraw(): Boolean {
        for (i in 0..8)
            if (listBoxOne[i] == ANSWER_NULL &&
                    listBoxTwo[i] == ANSWER_NULL)
                return false
        return true
    }

    override fun getEnd(): Boolean = end

    override fun stepBot(): Int {
        val checkStepWin = checkStepBot(listBoxTwo)
        return if (checkStepWin == ANSWER_NULL) {
            val checkStepEnemy = checkStepBot(listBoxOne)
            if (checkStepEnemy == ANSWER_NULL) {
                val list: MutableList<Int> = stepRandomBot()
                val random = Random(System.currentTimeMillis())
                list[random.nextInt(list.size)]
            } else
                checkStepEnemy
        } else
            checkStepWin
    }

    override fun getBoxWinList(): MutableList<Int> = boxWinList

    override fun setStep(boolean: Boolean) {
        step = boolean
    }

    override fun createLists() {
        boxWinList.clear()
        listBoxOne.clear()
        listBoxTwo.clear()
        for (i in 0..8) {
            listBoxOne.add(-1)
            listBoxTwo.add(-1)
        }
        end = false
    }

    override fun clearScore() {
        scorePlayerOne = 0
        scorePlayerTwo = 0
    }

    override fun setStepText(string: String) {
        stepText = string
    }

    private fun checkBox(list: MutableList<Int>): Boolean {
        for (i in 0..7) {
            var boxOne = false
            var boxTwo = false
            var boxThree = false

            if (AnswerField.getAnswer(i, 0) in list) boxOne = true
            if (AnswerField.getAnswer(i, 1) in list) boxTwo = true
            if (AnswerField.getAnswer(i, 2) in list) boxThree = true

            if (boxOne && boxTwo && boxThree) {
                boxWinList.add(AnswerField.getAnswer(i, 0))
                boxWinList.add(AnswerField.getAnswer(i, 1))
                boxWinList.add(AnswerField.getAnswer(i, 2))
                return true
            }
        }
        return false
    }

    private fun checkStepBot(list: MutableList<Int>): Int {
        for (i in 0..7) {
            var boxOne = false
            var boxTwo = false
            var boxThree = false

            if (AnswerField.getAnswer(i, 0) in list) boxOne = true
            if (AnswerField.getAnswer(i, 1) in list) boxTwo = true
            if (AnswerField.getAnswer(i, 2) in list) boxThree = true

            val answerOne: Int = AnswerField.getAnswer(i, 2)
            val answerTwo: Int = AnswerField.getAnswer(i, 1)
            val answerThree: Int = AnswerField.getAnswer(i, 0)

            if (boxOne && boxTwo &&
                    listBoxOne[answerOne] == ANSWER_NULL &&
                    listBoxTwo[answerOne] == ANSWER_NULL)
                return answerOne
            if (boxOne && boxThree &&
                    listBoxOne[answerTwo] == ANSWER_NULL &&
                    listBoxTwo[answerTwo] == ANSWER_NULL)
                return answerTwo
            if (boxTwo && boxThree &&
                    listBoxOne[answerThree] == ANSWER_NULL &&
                    listBoxTwo[answerThree] == ANSWER_NULL)
                return answerThree
        }
        return -1
    }

    private fun stepRandomBot(): MutableList<Int> {
        val list: MutableList<Int> = arrayListOf()
        for (i in 0..7) {
            val answerOne = AnswerField.getAnswer(i, 0)
            val answerTwo = AnswerField.getAnswer(i, 1)
            val answerThree = AnswerField.getAnswer(i, 2)
            for (item in listBoxOne) {
                if (answerOne != item &&
                        listBoxOne[answerOne] == ANSWER_NULL &&
                        listBoxTwo[answerOne] == ANSWER_NULL)
                    list.add(answerOne)
                if (answerTwo != item &&
                        listBoxOne[answerTwo] == ANSWER_NULL &&
                        listBoxTwo[answerTwo] == ANSWER_NULL)
                    list.add(answerTwo)
                if (answerThree != item &&
                        listBoxOne[answerThree] == ANSWER_NULL &&
                        listBoxTwo[answerThree] == ANSWER_NULL)
                    list.add(answerThree)
            }
        }
        return list
    }
}